crates/synthesis/src/multi_controlled/mcx.rs

```diff
@@ -0,0 +1,438 @@
+    circuit.h(4);
+    circuit.cp(PI2, 3, 4);
+    circuit.h(4);
+    circuit.compose(&rc3x()?, &[Qubit(0), Qubit(1), Qubit(2), Qubit(3)], &[])?;
+    circuit.h(4);
+    circuit.cp(-PI2, 3, 4);
+    circuit.h(4);
+    circuit.compose(
+        &rc3x()?.inverse()?,
+        &[Qubit(0), Qubit(1), Qubit(2), Qubit(3)],
+        &[],
+    )?;
+    circuit.compose(&c3sx()?, &[Qubit(0), Qubit(1), Qubit(2), Qubit(4)], &[])?;
+    Ok(circuit)
+}
+
+/// A block in the `action part`, see Iten et al.
+fn action_gadget() -> PyResult<CircuitData> {
+    let mut circuit = CircuitData::with_capacity(3, 0, 5, Param::Float(0.0))?;
+    circuit.h(2);
+    circuit.t(2);
+    circuit.cx(0, 2);
+    circuit.tdg(2);
+    circuit.cx(1, 2);
+    Ok(circuit)
+}
+
+/// A block in the `reset part`, see Iten et al.
+fn reset_gadget() -> PyResult<CircuitData> {
+    let mut circuit = CircuitData::with_capacity(3, 0, 5, Param::Float(0.0))?;
+    circuit.cx(1, 2);
+    circuit.t(2);
+    circuit.cx(0, 2);
+    circuit.tdg(2);
+    circuit.h(2);
+    Ok(circuit)
+}
+
+/// Synthesize a multi-controlled X gate with :math:`k` controls based on the paper
+/// by Iten et al. [1].
+///
+/// For :math:`k\ge 4` the method uses :math:`k - 2` dirty ancillary qubits, producing a circuit
+/// with :math:`2 * k - 1` qubits and at most :math:`8 * k - 6` CX gates. For :math:`k\le 3`
+/// explicit efficient circuits are used instead.
+///
+/// # Arguments
+/// - num_controls: the number of control qubits.
+/// - relative_phase: when set to `true`, the method applies the optimized multi-controlled
+///   X gate up to a relative phase, in a way that the relative phases of the `action part`
+///   cancel out with the relative phases of the `reset part`.
+/// - action_only: when set to `true`, the methods applies only the `action part`.
+///
+/// # References
+///
+/// 1. Iten et al., *Quantum Circuits for Isometries*, Phys. Rev. A 93, 032318 (2016),
+/// [arXiv:1501.06911] (http://arxiv.org/abs/1501.06911).
+pub fn synth_mcx_n_dirty_i15(
+    num_controls: usize,
+    relative_phase: bool,
+    action_only: bool,
+) -> PyResult<CircuitData> {
+    if num_controls == 1 {
+        let mut circuit = CircuitData::with_capacity(2, 0, 1, Param::Float(0.0))?;
+        circuit.cx(0, 1);
+        Ok(circuit)
+    } else if num_controls == 2 {
+        ccx()
+    } else if num_controls == 3 && !relative_phase {
+        c3x()
+    } else {
+        let num_ancillas = num_controls - 2;
+        let num_qubits = num_controls + 1 + num_ancillas;
+        let mut circuit = CircuitData::with_capacity(num_qubits as u32, 0, 0, Param::Float(0.0))?;
+
+        let controls: Vec<u32> = (0..num_controls).map(|q| q as u32).collect();
+        let target = num_controls as u32;
+        let ancillas: Vec<u32> = ((num_controls + 1)..num_qubits).map(|q| q as u32).collect();
+
+        for j in 0..2 {
+            if !relative_phase {
+                circuit.compose(
+                    &ccx()?,
```

Comment:
This seems a bit overcomplicated, wouldn't it be nicer to just define `circuit.ccx` as method? This would also incur at less overhead since you don't have to fetch the definitions etc.

Reply:
That's a good point. Since `ccx` and `rccx` are very common subcircuits, it does make perfect sense to add the corresponding methods to the ever-growing `CircuitDataForSynthesis` trait. I do prefer however to add _the definitions_ for CCX and RCCX rather than the gates themselves; while both would be valid options, the current way avoids an extra recursion step in `HighLevelSynthesis`.